base/dict.jl

```diff
@@ -352,17 +358,16 @@ function ht_keyindex2!(h::Dict{K,V}, key) where V where K
     return ht_keyindex2!(h, key)
 end
 
-@propagate_inbounds function _setindex!(h::Dict, v, key, index)
-    h.slots[index] = 0x1
-    h.keys[index] = key
-    h.vals[index] = v
+@propagate_inbounds function _setindex!(h::Dict{K,V}, v, key, index, sh) where {K, V}
+    h.slots[index] = sh
+    h.pairs[index] = Pair{K,V}(key, v)
     h.count += 1
     h.age += 1
     if index < h.idxfloor
         h.idxfloor = index
     end
 
-    sz = length(h.keys)
+    sz = length(h.pairs)
     # Rehash now if necessary
     if h.ndel >= ((3*sz)>>2) || h.count*3 > sz*2
```

Comment:
Thank you for the question. Generally yes, but here the primary motivation is different. The PR limits the number of `isequal` calls, and thus limits the number of allocations for abstract types and pressure on GC. These coeficients should be updated to limit memory consumption. This fine-tuning needs much more benchmarking on various sizes. There will always be some tradeoff between speed and used memory. I propose to move such a discussion into a separate PR. Meanwhile, I'll try to prepare some microbenchmarks.

Reply:
What I've tested so far, we would need to use SIMD (as in [DataStructures.jl](https://github.com/JuliaCollections/DataStructures.jl)) to increase the capacity without a significant performance drop (given by unpredictable branching). The good news is, that metadata in `slots` is prepared for that. Nevertheless, I'm not sure, if such a low-level llvm code should go to base because it will be harder to read and check ... and it will be platform specific.